Matter_Engine/Query/VolumetricMaterialTakeoff.cs

```diff
@@ -0,0 +1,123 @@
+            if (elementM == null)
+            {
+                Base.Compute.RecordError("Cannot query the VolumetricMaterialTakeoff from a null element.");
+                return null;
+            }
+
+            //If asked to look for fragment, and if fragment exists, return it
+            VolumetricMaterialTakeoff matTakeoff;
+            if (TryGetVolumetricMaterialTakeoffFragment(elementM, checkForTakeoffFragment, out matTakeoff))
+                return matTakeoff;
+
+            //IElementMs should implement one of the following:
+            // -SolidVolume and MaterialComposition or
+            // -VolumetricMaterialTakeoff
+            //This method first checks if the VolumetricMaterialTakeoff method can be found and run, and if so uses it.
+            //If not, it falls back to running the MaterialComposition and SolidVolume methods and gets the VolumetricMaterialTakeoff from them.
+
+
+            if (TryGetVolumetricMaterialTakeoff(elementM, out matTakeoff))
+                return matTakeoff;
+            else
+            {
+                MaterialComposition matComp;
+                double volume;
+                if (TryGetMaterialComposition(elementM, out matComp) && TryGetSolidVolume(elementM, out volume))
+                    return Create.VolumetricMaterialTakeoff(matComp, volume);
+                else
+                {
+                    Base.Compute.RecordError($"The provided element of type {elementM.GetType()} does not implement VolumetricMaterialTakeoff or MaterialComposition and SolidVolume methods. The VolumetricMaterialTakeoff could not be extracted.");
+                    return null;
+                }
+            }
+        }
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+
+        [Description("Tries running the VolumetricMaterialTakeoff method on the IElementM. Returns true if the method successfully can be found.")]
+        private static bool TryGetVolumetricMaterialTakeoff(this IElementM elementM, out VolumetricMaterialTakeoff volumetricMaterialTakeoff)
+        {
+            object result;
+            bool success = Base.Compute.TryRunExtensionMethod(elementM, "VolumetricMaterialTakeoff", out result);
+            volumetricMaterialTakeoff = result as VolumetricMaterialTakeoff;
+            return success;
+        }
+
+        /***************************************************/
+
+        [Description("Tries to find a VolumetricMaterialTakeoff attached as a fragment to the elementM. If found, returns true.")]
+        private static bool TryGetVolumetricMaterialTakeoffFragment(this IElementM elementM, bool checkForFragment, out VolumetricMaterialTakeoff takeoffFragment)
+        {
+            takeoffFragment = null;
+
+            //If bool is false, simply return false
+            if (!checkForFragment)
+                return false;
```

Comment:
Not sure I like this logic: a method should not have a boolean input asking whether it should to its job or not.

Reply:
More happy if I move the check up to the main methods?

Reply:
Yes definitely